src/python/grpcio/grpc/_cython/_cygrpc/time.pyx.pxi

```diff
@@ -12,14 +12,16 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from libc.math cimport modf as _cmodf, fabs as _cfabs
 
 cdef gpr_timespec _timespec_from_time(object time):
   cdef gpr_timespec timespec
+  cdef double whole_part
   if time is None:
     return gpr_inf_future(GPR_CLOCK_REALTIME)
   else:
-    timespec.seconds = time
-    timespec.nanoseconds = (time - float(timespec.seconds)) * 1e9
+    timespec.nanoseconds = <int32_t>(_cfabs(_cmodf(time, &whole_part)) * 1e9)
```

Comment:
Can we use `gpr_time_from_nanos` C API to perform the conversion?

Also, `modf` behaves like splitting -3.14 into -3 and -0.14. If you only take the absolute value of the fragment part, the represented time is changed from (-3.14 to -2.86 in the example).

Reply:
Is that the way C-core treats nanoseconds?

Reply:
Please check the `to_seconds_from_sub_second_time` function:
https://github.com/grpc/grpc/blob/0d9e6d1e363846066b79dc41416ee374ca1657d1/src/core/lib/gpr/time.cc#L82

Seems our team does spent a lot of time to design negative time representation.

Reply:
I think we can simply rely on `gpr_time_from_nanos` and `gpr_micros_from_time` to convert directly from a `double`. I don't know how best to test this, as cygrpc symbols are not visible.

Reply:
How about testing if negative deadline for RPC `timeout` or Completion Queue `poll` will result in crashing the whole process?

Reply:
We cannot use the public interface because `timeout` is not an absolute point in time, but is added to current timestamp. We could potentially do that in cython direct channel test. Let me investigate.